gossipd/gossip.c

```diff
@@ -1634,37 +1638,32 @@ static struct io_plan *connection_out(struct io_conn *conn,
 				   handshake_out_success, reach);
 }
 
-static void try_connect(struct reaching *reach);
-
 static void connect_failed(struct io_conn *conn, struct reaching *reach)
 {
-	u32 diff = time_now().ts.tv_sec - reach->first_attempt;
-	reach->attempts++;
-
-	if (!important_peerid_map_get(&reach->daemon->important_peerids, &reach->id)
-	    && reach->attempts >= 10) {
-		status_info("Failed to connect after %d attempts, giving up "
-			    "after %d seconds",
-			    reach->attempts, diff);
-		daemon_conn_send(
-		    &reach->daemon->master,
-		    take(towire_gossip_peer_connection_failed(
-			NULL, &reach->id, diff, reach->attempts, false)));
-		tal_free(reach);
-	} else {
-		unsigned int secs;
-
-		/* Exponential backoff, then every 5 minutes */
-		if (reach->attempts < 9)
-			secs = 1 << reach->attempts;
-		else
-			secs = 300;
-		status_trace("Failed connected out for %s, will try again in %u seconds",
-			     type_to_string(tmpctx, struct pubkey, &reach->id),
-			     secs);
-		new_reltimer(&reach->daemon->timers, reach, time_from_sec(secs),
-			     try_connect, reach);
+	u8 *msg;
+	struct important_peerid *imp;
+
+	/* Tell any connect commands what happened. */
+	msg = towire_gossipctl_connect_to_peer_result(reach, &reach->id,
+						      false, strerror(errno));
+	for (size_t i = 0; i < reach->num_master_responses; i++)
+		daemon_conn_send(&reach->daemon->master, msg);
+
+	status_trace("Failed connected out for %s",
+		     type_to_string(tmpctx, struct pubkey, &reach->id));
+
+	/* If we want to keep trying, do so. */
+	imp = important_peerid_map_get(&reach->daemon->important_peerids,
+				       &reach->id);
+	if (imp) {
+		/* FIXME: Exponential backoff! */
+		status_trace("...will try again in %u seconds", 5);
+		/* If important_id freed, this will be removed too */
+		new_reltimer(&reach->daemon->timers, imp,
```

Comment:
Should we not keep the returned `struct oneshot` in the `struct important_peerid`?  And before making a new timer, we should free any existing timers?  Otherwise `connect` commands may cause multiple overlapping timers to exist (I think?).

    /* Kick the timer if present, or start a new one if not present. */
    tal_free(imp->timer);
    imp->timer = new_reltimer(&reach->daemon->timers, imp, time_from_sec(5), retry_important, imp);

The `retry_important` function also needs to clear this timer:

    imp->timer = tal_free(imp->timer);

Reply:
Good point, we will end up with one timer per connect for every important peer.  Will fix, though not here.